### ISSUE TITLE
Bizarre numpy recarray string dimension mismatch bug

### ISSUE THREAD

**Issue by synapticarbors** [cython/cython#1689]
I ran into a very bizarre bug in some of my work where I pass a numpy recarray to a cython function that contains a string field. For some reason if the string field length contains a 9 and is greater than 9 (e.g 19, 29, 39, 49, 191, etc), I get an error that looks like:

```
ValueError: Expected a dimension of size 39, got 9
```

There are a few exceptions. A length of 91 seems to be ok, but 919 throws an error. Here is a minimal example that demonstrates the issue with Cython 0.25.2:

```
import numpy as np
cimport numpy as np


dtype = [
         ('f1', 'S39')
        ]

a = np.recarray(1, dtype=dtype)
a[0] = tuple(['abcd',])


cdef packed struct r_type:
    char[39] f1


def test_struct(r_type[:] x):
    return x[0].f1
    
print test_struct(a)
```

Any idea what could possibly be going on here?

**Comment by synapticarbors** [comment]
I went back and tested versions of cython going back to 0.19 and this bug seems to be present in all of them as far as I can tell.

**Comment by synapticarbors** [comment]
Recently re-looked at this issue and tested for a bunch of more recent versions of Cython since I originally reported this. Scanning through versions, it appeared as if this was fixed between 0.29.5 (failed) and 0.29.6 (correct). The Changelog indicates that this was fixed while dealing with https://github.com/cython/cython/issues/2845. It appears to be the same issue.

Thank you @scoder for fixing this one.
